### DISCUSSION TITLE
FastAPI application is processing same request multiple times in Backend

### DISCUSSION THREAD

**Discussion by rksingh53** [fastapi/fastapi#8369] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
###This API takes id in request , creates temp path in container , searches path for this id in database and copies file from AWS s3b for this id in temp created path, does ML processing and deletes the temp path then returns the predicted data

from datetime import datetime
import uvicorn
from fastapi import Depends, FastAPI, HTTPException, Request
from starlette.middleware.cors import CORSMiddleware
# initializing FastAPI
from fastapi import FastAPI, Response, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel
from fastapi import Security ,HTTPException
from fastapi.security.api_key import APIKeyHeader
import pathlib
temp = pathlib.PosixPath
#pathlib.PosixPath = pathlib.WindowsPath

import torch
# Logging
import logging, logging.config

logging.config.dictConfig(
    {
        'version': 1,
        'disable_existing_loggers': True,
    }
)
global logger
# logging.basicConfig(filename='app.log', filemode='w',format='%(asctime)s:%(levelname)s:%(message)s', level=logging.DEBUG)

app = FastAPI(docs_url="/", redoc_url=None,
              title="First ML API",
              description="First ML API",
              version="V 1.0",
              )
# FileDownload()
# SMBConnectionDownload()
config = readconfig()

API_KEY = config.API_KEY
API_KEY_NAME = config.API_KEY_NAME

api_key_header_auth = APIKeyHeader(name=API_KEY_NAME, auto_error=True)

async def get_api_key(api_key_header: str = Security(api_key_header_auth)):
    if api_key_header != API_KEY:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid API Key",
        )


origins = [
    "http://localhost.tiangolo.com",
    "https://localhost.tiangolo.com",
    "http://localhost",
    "http://localhost:8080",
    "http://localhost:4200",
]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)



class Item(BaseModel):
    id : str



logging.config.dictConfig(
    {
        'version': 1,
        'disable_existing_loggers': True,
    }
)
global logger

logger = logging.getLogger("main")
logger.setLevel(logging.INFO)
# create the logging file handler
fh = logging.FileHandler("app.log")
ch = logging.StreamHandler()
ch.setLevel(logging.DEBUG)
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
fh.setFormatter(formatter)
ch.setFormatter(formatter)
while logger.handlers:
    logger.handlers.pop()
# add handler to logger object
logger.addHandler(fh)
logger.addHandler(ch)
logger.propagate = False



@app.get("/")
async def root():
    return {"message": "Hello Bigger Applications!"}


@app.get("/heartbeat")
def heartbeat():
    return "Success"


@app.get('/predictdata',dependencies=[Security(get_api_key)])
def predictdata(id: str):


    try:
        res = {"id" : id , "prediction": "" , "StatusCode" : "200", "Validations" : []}
        
        if not id.isdigit():
            res["Validations"].append({"Type" : "Validation", "Description" : "id must be a integer"})
            raise TypeError("id must be an integer")
        
        logger.info("*** ML prediction API started for Id : {} ***".format(id))
        logger.info("########################")
        data = ML_Process(id, res)
        res = data
        logger.info(res)

        return JSONResponse(status_code=int(res["StatusCode"]), content=res)

    except Exception as ex:


            traceback.print_exc()
            error_trace = traceback.format_exc()
            res["Validations"].append({"Type" : "Error","Description" : error_trace})
            res["StatusCode"] = str(500)
            logger.exception("Exception in my ML API")
            logger.error(ex,exc_info=True)

    return JSONResponse(status_code=500, content=res)

	
def ML_Process(id, res):
    #get docpath from db by id , copy doc to temp path from s3b and run ML model on the doc
    #after processing it deletes temp path and returns predicted data
    # this is a long running process takes avg 25 second but can take sometime 1-2 minutes.
    # Issue is observed only when it's a long running requests e.g Model takes more time to do inference
  


if __name__ == "__main__":
    uvicorn.run(
        app,
        host="0.0.0.0",
        port=8422
    )


-------------------------------------

Dockerfile

FROM python:3

WORKDIR /app

COPY requirements.txt requirements.txt

RUN pip install -r requirements.txt


RUN apt-get update

RUN apt-get install poppler-utils -y && \
    apt install ghostscript -y && \
    apt-get install -y tesseract-ocr && \


COPY . .

WORKDIR /app

CMD uvicorn app:app --host 0.0.0.0 --port 5057 



-------------------------------------------


Kubernetes pod logs of application which shows same request getting processed multiple times

i sent the request of id - 12 only once through the swagger  and nobody else was using it 

log :-

2022-08-11 04:27:22,070 - main - INFO - *** ML prediction API started for id : 12 ***
2022-08-11 04:27:22,071 - main - INFO - ########################
2022-08-11 04:27:22,072 - main - INFO - Temp folder structure created :  /app/12_20220811042722071864
2022-08-11 04:27:22,104 - main - INFO - File existing in this id folder in S3B folder  .
2022-08-11 04:27:22,105 - main - INFO - File existing in this id folder in S3B folder  ..
2022-08-11 04:27:22,105 - main - INFO - File existing in this id folder in S3B folder  278692642.pdf
2022-08-11 04:27:22,591 - main - INFO - "278692642.pdf" file copied in temp directory
2022-08-11 04:27:27,518 - main - INFO - 278692642.pdf
2022-08-11 04:27:27,706 - main - INFO - /app/12_20220811042722071864/278692642.pdf
2022-08-11 04:28:56,711 - main - INFO - prediction  =  done
2022-08-11 04:28:56,711 - main - INFO - -----------------------------------------------------------------------------
2022-08-11 04:28:22,078 - main - INFO - *** ML prediction API started for id : 12 ***
2022-08-11 04:28:22,080 - main - INFO - ########################
2022-08-11 04:28:22,080 - main - INFO - Temp folder structure created :  /app/12_20220811042822080625
2022-08-11 04:28:22,292 - main - INFO - File existing in this id folder in S3B folder  .
2022-08-11 04:28:22,292 - main - INFO - File existing in this id folder in S3B folder  ..
2022-08-11 04:28:22,292 - main - INFO - File existing in this id folder in S3B folder  278692642.pdf
2022-08-11 04:28:24,784 - main - INFO - "278692642.pdf" file copied in temp directory
2022-08-11 04:28:38,689 - main - INFO - 278692642.pdf
2022-08-11 04:28:39,986 - main - INFO - /app/12_20220811042822080625/278692642.pdf
2022-08-11 04:28:56,711 - main - INFO - prediction  =  done
2022-08-11 04:28:56,711 - main - INFO - -----------------------------------------------------------------------------
█

 |----------------------------------------| 0.00% [0/1 00:00<?]

 |████████████████████████████████████████| 100.00% [1/1 00:16<00:00]

                                                                     

                                                                     
█

 |----------------------------------------| 0.00% [0/1 00:00<?]

 |████████████████████████████████████████| 100.00% [1/1 00:14<00:00]

                                                                     

2022-08-11 04:29:22,145 - main - INFO - Deleted Temp folder : /app/12_20220811042722071864
2022-08-11 04:29:22,145 - main - INFO - ML prediction Done!!!
2022-08-11 04:29:22,145 - main - INFO - {"id" : 12,'prediction': 'predicted value here', 'StatusCode': '200', 'Validations': []}       
2022-08-11 04:29:22,077 - main - INFO - *** ML prediction API started for id : 12 ***
2022-08-11 04:29:22,077 - main - INFO - ########################
2022-08-11 04:29:22,077 - main - INFO - Temp folder structure created :  /app/12_20220811042922077785
2022-08-11 04:29:22,100 - main - INFO - File existing in this id folder in S3B folder  .
2022-08-11 04:29:22,101 - main - INFO - File existing in this id folder in S3B folder  ..
2022-08-11 04:29:22,101 - main - INFO - File existing in this id folder in S3B folder  278692642.pdf
2022-08-11 04:29:22,145 - main - INFO - Deleted Temp folder : /app/12_20220811042822080625
2022-08-11 04:29:22,145 - main - INFO - ML prediction Done!!!
2022-08-11 04:29:22,145 - main - INFO - {"id" : 12,'prediction': 'predicted value here', 'StatusCode': '200', 'Validations': []}
2022-08-11 04:28:38,689 - main - INFO - 278692642.pdf
2022-08-11 04:28:39,986 - main - INFO - /app/12_20220811042922077785/278692642.pdf
2022-08-11 04:28:56,711 - main - INFO - prediction  =  done
2022-08-11 04:28:56,711 - main - INFO - -----------------------------------------------------------------------------

----------------------| 0.00% [0/1 00:00<?]

 |████████████████████████████████████████| 100.00% [1/1 00:00<00:00]

                                                                     

2022-08-11 04:29:40,316 - main - INFO - Deleted Temp folder : /app/12_20220811042922077785
2022-08-11 04:29:40,316 - main - INFO - ML prediction Done!!!
2022-08-11 04:29:40,318 - main - INFO - {"id" : 12,'prediction': 'predicted value here', 'StatusCode': '200', 'Validations': []}


---- 

here for first time, it took the request id started processing but before completing i.e before deleting temp path, second process/thread is initiated as i see this log again 
 **ML prediction API started for id : 12 
and a new temp path is also created
2022-08-11 04:28:22,080 - main - INFO - Temp folder structure created :  /app/12_20220811042822080625**
and starts processes same request again before first one is complete then again after few seconds third process/thread is initiated and starts processing on same request. it shouldn't process same request again and again. I don't have any multiprocessing or multithreading concept then why its processing same request again and again. plz guide
```

### Description

I don't have any multiprocessing or multithreading concept in my Fast API Application then why its processing same request again and again.  Its a simple ML Processing Application

### Operating System

Linux

### FastAPI Version

fastapi-0.68.2

### Python Version

3.8.5

**Comment by JarroVGIT** [comment]
This is just a wall of code and text. Please provide a minimal reproducible example that shows the unwanted/unexpected behaviour. Nobody can reproduce this properly without all the auxiliaries in place (like K8S, etc). 

Did you by any chance double click on the 'execute' button in SwaggerUI?

**Comment by rksingh53** [comment]
Sure .. could you please let me know what details are required? I have provided the API code and Dockerfile I'm using . Plz let me know which other code i should be sharing here

**Comment by JarroVGIT** [comment]
My point is that it is impossible for others to reproduce this. Try to minimise the code necessary to reproduce this. For example; is this behaviour only in K8S? Or also when you run the container locally? If it is also locally, then we can start ruling out K8S interfering behaviours. 

Given the code you provided, I see no immediate cause for it to be called twice. I do see double `global logger` and some double config going on, but I don't think that should cause this behaviour. 

Is the behaviour consistent? Is it **always** ran twice when called once? Or is it sometimes? Do you have a load balancer in place that could be misconfigured (e.g. are we looking at logs from the Deployment rather than the individual pod and are requests sent to all pods in the deployment?) I am just throwing options here ;)

**Comment by rksingh53** [comment]
> My point is that it is impossible for others to reproduce this. Try to minimise the code necessary to reproduce this. For example; is this behaviour only in K8S? Or also when you run the container locally? If it is also locally, then we can start ruling out K8S interfering behaviours. 
> 
> Given the code you provided, I see no immediate cause for it to be called twice. I do see double `global logger` and some double config going on, but I don't think that should cause this behaviour. 

This issue only happens when the application is deployed on k8s . When I ran same application on my local windows machine using Docker Desktop, this issue is not reproducible

**Comment by JarroVGIT** [comment]
Then I would start tracking requests through your K8S stack somehow. Maybe some component is awaiting the request response and thinks that >20 seconds responses take too long and it resends the request?

**Comment by rksingh53** [comment]
> Then I would start tracking requests through your K8S stack somehow. Maybe some component is awaiting the request response and thinks that >20 seconds responses take too long and it resends the request? 

Could you please look into my API declaration part and route declaration once and see if I have used correctly.

I have checked in pod when such issues happen, there are multiple threads starting multiple processes for same request and I am unable to find out why this is happening . Because in my uvicorn I'm not defining multiple threads also I don't have used async in my route function,it's simple def function.

Logging isn't the issues because even when I removed the logs completely still issue was happening.


Question :- Do we need to define our API in a specific way if our application takes long time for request processing if we are to deploy it on k8s ?

**Comment by rksingh53** [comment]
This issue is intermittent mostly but it's always reproducible when I scale down the pods resources i e when I bring down my pod's cpu and memory limits.

I can sense the issue may be related to this because with less Infra , ML model takes longer to Inference  (we know Torch library needs more cpu cores to run smoothly ) and same request starts getting processed multiple times by multiple threads/processes but I am not sure if this is the only reason.  Also if it is , kindly suggest any way I can avoid duplicate processing of same request from code?

**Comment by JarroVGIT** [comment, marked as answer]
> Could you please look into my API declaration part and route declaration once and see if I have used correctly.

I have and I see nothing wrong with it. To me, it seems that something (a component in your stack) is retrying. There are many components in play when deployed to K8S and it is impossible to say where it could be (because it could be anyone of them) as we don't know your full stack. All I can say that it is not FastAPI that is causing this, that part is doing its job properly. 

> This issue is intermittent mostly but it's always reproducible when I scale down the pods resources i e when I bring down my pod's cpu and memory limits.

This indicates (but not proves!) that the component that is sending the request to Uvicorn (your container) retries when the container is moved to another resource (node). If you scale up/down the requested resources of a pod, the pod might be rescheduled somewhere else where those resources are available. Again, I am just guessing here as I cannot see what you are doing, how your cluster looks like and what components you have in your entire stack. I hope this gives at least some pointers on what to look for.  

As final point: I don't see any Uvicorn logs in your logs. You could try to find out what is sending the request (is Uvicorn receiving the request twice, or is Uvicorn the part that is sending the request twice (the latter, I don't think so but best to be sure)).

**Comment by JarroVGIT** [comment]
Hi @rksingh53 , have you found the culprit? I am curious what was causing this for you, it's an interesting case!

**Comment by rksingh53** [comment]
> Hi @rksingh53 , have you found the culprit? I am curious what was causing this for you, it's an interesting case!

No, I tried a lot finding root cause for this but couldn't get anything..

**Comment by cglacet** [comment]
I have the same issue, I found a very simple example that triggers the behaviour. It feel like there is some kind of hidden 60 seconds "retry timeout" (triggers any endpoint function to re-execute if it takes more than 60 seconds). I searched the documentation and the codebase for it but without success.

## My config

- Minikube `v1.28.0`
- Kubernetes `v1.25.2`
- Docker image `python:3.10-slim-buster`
- fastapi `0.95.0`
- starlette `0.26.1`

My liveness and startup probes are as follow:

```yaml
livenessProbe:
  httpGet:
    path: /healthz
    port: {{ .Values.app.port }}
  failureThreshold: 3
  initialDelaySeconds: 60
  periodSeconds: 3
startupProbe:
  httpGet:
    path: /healthz
    port: {{ .Values.app.port }}
  failureThreshold: 30
  periodSeconds: 10
```

## How to reproduce 

```python
import uuid
import logging 
from typing import Dict

from fastapi import FastAPI
from pydantic import BaseModel


app = FastAPI()
logging.basicConfig(
    format='%(asctime)s %(levelname)-8s %(message)s',
    level=logging.INFO,
    datefmt='%H:%M:%S',
)
logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)


class SleepDurationInput(BaseModel):
    seconds: int


@app.get("/sleep")
async def sleep(input: SleepDurationInput) -> Dict[str, str]:
    request_id = uuid.uuid4()
    logger.debug("[%s] Start sleeping for %s seconds", request_id, input.seconds)
    await asyncio.sleep(input.seconds)
    logger.debug("[%s] Done  sleeping for %s seconds", request_id, input.seconds)
    return { "info": f"slept for {input.seconds} seconds" }


@app.get("/healthz")
async def health_check():
    logging.debug("K8s health check")
    return "ok"
```

If the sleep duration is small (for example 5 seconds), it works. But if you put a longer sleep duration (> 60 sec.) if does run the code several times.

With 100 seconds it gives the following output after a single call to the `/sleep` endpoint:

```python
2023-05-15T09:45:56.169Z | DEBUG:app:[e3388592-1cd7-40f6-a6f7-53e79692cfe0] Start sleeping for 100 seconds
2023-05-15T09:45:57.115Z | INFO:     172.17.0.1:38564 - "GET /healthz HTTP/1.1" 200 OK
...
2023-05-15T09:46:54.114Z | INFO:     172.17.0.1:39512 - "GET /healthz HTTP/1.1" 200 OK
2023-05-15T09:46:56.169Z | DEBUG:app:[da1c1fdb-cbad-40f4-8cfd-01eacb9a88cb] Start sleeping for 100 seconds
```

As you can see, after 60 seconds, the `sleep` function is called a second time.

**Comment by notypecheck** [comment]
`proxy-read-timeout` in nginx ingress controller is set to 60 seconds by default

https://docs.nginx.com/nginx-ingress-controller/configuration/global-configuration/configmap-resource/

**Reply by cglacet** [reply]
Adding the folowing `nginx.ingress.kubernetes.io/proxy-read-timeout` annotations to my ingress config fixed it:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  ...
  annotations:
    kubernetes.io/ingress.class: nginx
    nginx.ingress.kubernetes.io/proxy-read-timeout: "180"
```

## [proxy-read-timeout](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/#proxy-read-timeout)
Sets the timeout in seconds for [reading a response from the proxied server](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout). The timeout is set only between two successive read operations, not for the transmission of the whole response.

**Comment by dss010101** [comment]
I am encountering this same issue, but with Azure ContainerApps.  Havent been able to figure out as in my case it also is not reproducible locally.  In my case i have 4 workers.  The code am running does parallel processing using either ayncio,gather or a Threadpool.   Some workers are configured for one vs the other for testing performance between the two methods of parallelizing queries.  Both hit a postgres database and can create as much as up to 40 connections in a given process using connection pools.
the asyncio.gather code using asyncpg library.  while the Threading uses psycopg3 synchronous methods.

asyncio code looks like this:
```
pool = await asyncpg.create_pool(**self.connect_str_asyncpg)
         async with pool.acquire() as conn:
            tasks = [pool.fetch(sql, *p) for p in params]
            completed = await asyncio.gather(*tasks, return_exceptions=True)
         
```

and the ThreadPool code looks like this:
```
          loop = asyncio.get_running_loop()
         with ThreadPoolExecutor() as pool:            
            futures = { loop.run_in_executor(pool, self.read_sql_sync, sql, p): p for  p in params }  
            completed, _ = await asyncio.wait(futures.keys(), return_when=asyncio.ALL_COMPLETED)
````

What i have observed so far is...it only happens (randomly seemingly) when the work starts with a worker using ThreadPool.  The work will complete, but before the data can be returned to the client, FastApi resubmits the request to a worker that is configured for using the asyncio.gather instead of a ThreadPool (have no idea why or how it knows which is which..but its always that way. maybe at the time this happens, all the workers using Threadpool are seen has in a bad state by FastApi?).

The other observation is that i noticed the connections returned by the log output below is normally around 22. but just before this issue occurs, its around 28/29. 

```
@app.middleware("http")
async def track_requests(request: Request, call_next):
    req_id = str(uuid.uuid4())
    request_id.set(req_id)

    # Check file descriptors
    soft, hard = resource.getrlimit(resource.RLIMIT_NOFILE)
    current_fds = len(psutil.Process().open_files())

    config.logger.info(f"Worker {os.getpid()} starting request {req_id}. FDs: {current_fds}/{soft}, Connections: {len(psutil.Process().connections())}")    
    try:
        response = await call_next(request)
        config.logger.info(f"Worker {os.getpid()} completed request {req_id}")
        return response
    except Exception as e:
        config.logger.error(f"Worker {os.getpid()} failed request {req_id}: {str(e)}")
        raise
````

So im here trying to understand what are the things FastAPI looks for to determine whether it needs to resubmit a request or not.
Any help or ideas appreciated

**Comment by DocBoo** [comment]
This seems not to be related with your environment. 
I've the same in local environment on Linux Mint 22/FastAPI

from fastapi import FastAPI
from fastapi.responses import HTMLResponse
from fastui import prebuilt_html

app = FastAPI()

@app.get('/{path:path}')
async def html_landing() -> HTMLResponse:
    return HTMLResponse(prebuilt_html(title='FastUI Demo'))

**Launch local server:** 
uvicorn main:app --reload

**Then in your brower (Chrome in may case):** 
http://127.0.0.1:8000/test

**Trace:**
INFO:     Application startup complete.
INFO:     127.0.0.1:52124 - "GET /test HTTP/1.1" 200 OK
INFO:     127.0.0.1:52124 - "GET /.well-known/appspecific/com.chrome.devtools.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:52124 - "GET /api/test HTTP/1.1" 200 OK
INFO:     127.0.0.1:52134 - "GET /favicon.ico HTTP/1.1" 200 OK
INFO:     127.0.0.1:52134 - "GET /api/test HTTP/1.1" 200 OK

# THEN

If I replace my function by:

@app.get('/{path:path}')
async def html_landing() -> HTMLResponse:
    return { 'message': 'Error: This endpoint is not implemented.' }

**Trace:**
INFO:     Application startup complete.
INFO:     127.0.0.1:33554 - "GET /test HTTP/1.1" 200 OK
INFO:     127.0.0.1:33554 - "GET /.well-known/appspecific/com.chrome.devtools.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:33554 - "GET /favicon.ico HTTP/1.1" 200 OK

==> No more problem

**Also works:**

@app.get('/{path:path}', response_class=HTMLResponse)
async def html_landing():
    return HTMLResponse(content="Error: This endpoint is not implemented.", status_code=200)
